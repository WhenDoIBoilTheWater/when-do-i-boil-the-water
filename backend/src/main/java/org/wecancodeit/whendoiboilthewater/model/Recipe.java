package org.wecancodeit.whendoiboilthewater.model;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;

import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.ManyToMany;
import javax.persistence.OneToMany;

import com.fasterxml.jackson.annotation.JsonIgnore;

@Entity
public class Recipe {
	@GeneratedValue @Id private Long id;
	private String name;	
	private int servingSize;
	private String description;
	private Long length;
	@OneToMany(mappedBy = "recipe")
	private List<Step> steps = new ArrayList<Step>();
	@JsonIgnore @ManyToMany
	private Collection<Ingredient> ingredients = new HashSet<Ingredient>();
	@JsonIgnore @ManyToMany
	private Collection<Meal> meals = new HashSet<Meal>();
	
	public Recipe() {}
	

	
	public Recipe( String name, int servingSize, String description) {
		this.name = name;
		this.servingSize = servingSize;
		this.description = description;
		this.length = calculateLength();
	}

	public Long getId() {
		return id;
	}

	public List<Step> getSteps() {
		return steps;
	}

	public Collection<Ingredient> getIngredients() {
		return ingredients;
	}

	public Long getLength() {
		return length;
	}

	public String getName() {
		return name;
	}

	public int getServingSize() {
		return servingSize;
	}

	public Collection<Meal> getMeals() {
		return meals;
	}

	public String getDescription() {
		return description;
	}

	
	//will we ever use this?
	public void addStep(String description, Long secBeforeEnd) {
		steps.add(new Step(secBeforeEnd, description));
		length = calculateLength();
	}
	
	public void addStep(Step step) {
		steps.add(step);
		length = calculateLength();
	}
	
	public void addIngredient(Ingredient ingredient) {
		ingredients.add(ingredient);
	}
	
	public Long calculateLength() {
		Long longestStepTime = 0L;
		for (Step step : steps) {
			if (step.getSecBeforeEnd() > longestStepTime) {
				longestStepTime = step.getSecBeforeEnd();
			}
		}
		return longestStepTime;
	}

}
