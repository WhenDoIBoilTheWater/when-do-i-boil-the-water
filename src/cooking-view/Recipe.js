import React from "react";
import TimerCard from "./TimerCard.js"
import CurrentStepCard from "./CurrentStepCard"

export class Recipe extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            recipe: this.props.recipe,
            meal: this.props.meal,
            arrayOfTimers: [],
            globalSeconds: 0,
            currentStepDescription: 'loading...'
        };

        this.buildTimers()

        this.tick = this.tick.bind(this);
    }   

    tick =()=>{
        this.state.arrayOfTimers.forEach(timer=>{
            if (timer.when === this.state.globalSeconds){
                timer.callback()
            }
        })
        this.setState({globalSeconds: this.state.globalSeconds+1})
    }

    setTimer(when, description, callback){
        const timer = {when, description, callback}
        this.state.arrayOfTimers.push(timer)
    }

    buildTimers(){
        this.state.recipe.steps.forEach((step)=>{
            this.setTimer(this.state.meal.length - step.secBeforeEnd, step.description, () => {
                //callback function:
                let length
                if(this.state.arrayOfTimers[1]){
                    length = this.state.arrayOfTimers[1].when - this.state.globalSeconds
                } else {
                    length = 0
                }
                this.setState({
                    currentStepDescription: step.description,
                    currentStepLength: length
                })
                this.state.arrayOfTimers.shift()
            })
        })
        setInterval(this.tick, 1000)

        this.setTimer(this.state.meal.length, 'Serve', '',() => {
            this.setState({
               currentStepRecipe: 'And finally',
               currentStepDescription: 'Serve!',
               currentStepLength: 0
           })

            this.state.arrayOfTimers.shift();
        })

        this.state.arrayOfTimers.sort(function (a, b) {
            return a.when - b.when;
        });
    }



    render() {
    	let currentStep
    	if (this.state.currentStepDescription !== 'loading...'){
            currentStep = <CurrentStepCard description={this.state.currentStepDescription} length={this.state.currentStepLength} />
        }

        return (
            <section>
            <h4>{this.state.globalSeconds}</h4>
            {currentStep}
            <ul className="list-of-steps">
            {this.state.arrayOfTimers.map(timer => {
                return(
                    (<TimerCard key={timer.description} when={timer.when} description={timer.description} />)
                    )
            })}
            </ul>

            </section>
            /*<SpotifyWidget />*/
            );
    }
}

export default Recipe