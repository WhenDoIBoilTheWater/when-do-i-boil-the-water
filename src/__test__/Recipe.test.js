import React from 'react';
import Recipe from '../cooking-view/Recipe.js';
import renderer from 'react-test-renderer';

it('renders nothing without given anything from planning-view', () => {
  const tree = renderer.create(
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
