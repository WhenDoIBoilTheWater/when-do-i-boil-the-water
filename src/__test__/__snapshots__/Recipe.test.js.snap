// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders nothing without given anything from planning-view 1`] = `null`;
